src/uct/ib/base/ib_device.c

```diff
@@ -482,18 +482,119 @@ void uct_ib_handle_async_event(uct_ib_device_t *dev, uct_ib_async_event_t *event
     ucs_log(level, "IB Async event on %s: %s", uct_ib_device_name(dev), event_info);
 }
 
+static ucs_status_t uct_ib_device_get_path_buffer(uct_ib_device_t *dev,
+                                                  char *path_buffer)
+{
+    char *resolved_path;
+
+    resolved_path = realpath(dev->ibv_context->device->ibdev_path, path_buffer);
+    if (resolved_path == NULL) {
+        return UCS_ERR_IO_ERROR;
+    }
+
+    /* Make sure there is "/infiniband/" substring in path_buffer */
+    if (strstr(path_buffer, "/infiniband/") == NULL) {
+        return UCS_ERR_IO_ERROR;
+    }
+
+    return UCS_OK;
+}
+
+char* uct_ib_device_get_pcie_bus(char *path_buffer, uint8_t layers)
+{
+    while (layers--) {
+        path_buffer = dirname(path_buffer);
+        if (path_buffer == NULL) {
+            return NULL;
+        }
+    }
+
+    return basename(path_buffer);
+}
+
+static ucs_status_t uct_ib_device_get_ids_from_path(char *path,
+                                                    long *vendor_id,
+                                                    long *device_id)
+{
+    ucs_status_t status;
+
+    strcat(path, "/%s");
+
+    status = ucs_read_file_number(vendor_id, 1, path, "vendor");
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    status = ucs_read_file_number(device_id, 1, path, "device");
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    return UCS_OK;
+}
+
+static ucs_status_t uct_ib_device_get_sf_ids(uct_ib_device_t *dev)
+{
+    long vendor_id, device_id;
+    char path_buffer[PATH_MAX];
+    char pci_bus[64];
+    char *pci_bus_end_ptr;
+    char *bus_id;
+    char ids_path[PATH_MAX];
+    ucs_status_t status;
+
+    /* realpath name is of form /sys/devices/.../0000:03:00.0/<UUID>/infiniband/mlx5_0
+     * and bus_id is constructed from 0000:03:00.0 */
+
+    status = uct_ib_device_get_path_buffer(dev, path_buffer);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    bus_id = uct_ib_device_get_pcie_bus(path_buffer, 3);
```

Comment:
seems like 'bus_id' here, after stripping 3 directory components, already points to the SF
```
$ cat $(realpath /sys/class/infiniband/mlx5_0)/../../../vendor
0x15b3
$ cat $(realpath /sys/class/infiniband/mlx5_0)/../../../device
0xa2d6
$ cat $(realpath /sys/class/infiniband/mlx5_0)/../../../current_link_speed
16 GT/s
```

Reply:
@yosefe Fixed.